### PR TITLE
[Frontend] Fix: modify randomNonce role

### LINKED ISSUE #269: [frontend] epoch key nonce is wrong
**Describe the bug**
Now the nonce is randomly chosen by
https://github.com/social-tw/social-tw-website/blob/3ff97adfda630ca3e847752d70a5aa21de29bf55/packages/frontend/src/utils/randomNonce.ts#L1-L3

There are two problems
1. There are `NUM_EPOCH_KEY_NONCE_PER_EPOCH=3` nonces
    But this function only random chooses `0` and `1`
    The outcome should be `0, 1, 2`
2. We should allow users to use 3 different epoch keys in their first 3 actions (discussed with Nicole)
    so it should query the `actionCount` from relay and set `0`, `1`, `2` for the first 3 actions (**not random**)
    after the 4th action, it random chooses nonce from `0`, `1`, `2` (**random**)

### PR DESCRIPTION
**Pull request by UranusLin**
## Summary

This PR updates the randomNonce function to correctly generate nonces based on the user's action count. It ensures that the first three nonces (0, 1, 2) are used in order for the initial three actions, and subsequent actions receive a randomly chosen nonce within the same range.

## Linked Issue

fix #269 

## Details

- Updated `randomNonce` to consider the action count before deciding whether to return a sequential nonce or a random nonce.
- Imported `NUM_EPOCH_KEY_NONCE_PER_EPOCH` from config to use as the upper limit for nonce generation.

## Impacted Areas

- `packages/frontend/src/utils/randomNonce.ts`
- `packages/frontend/src/__tests__/utils/randomNonce.test.ts`

## Tests

- Unit tests have been updated in randomNonce.test.ts to cover the new logic.
- Tests ensure the function returns sequential nonces for the first three actions and random nonces thereafter.
- No specific environment or configuration needed apart from the standard project setup.

## Possible Impacts

- This change ensures more predictable and correct nonce generation for the first three user actions, impacting how the application handles these initial transactions.
- No backward compatibility issues are expected.

## Visual Materials

N/A

## Verification Steps

- Run the updated unit tests in randomNonce.test.ts to verify the correct function behavior.
- Integrate the function into the frontend and perform manual testing to observe nonce generation based on action counts.

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

-   [x] Update randomNonce function logic.
-   [x] Update and verify unit tests for randomNonce.

## Checklist

-   [x] All new and existing tests pass
-   [x] I have updated the documentation (if applicable)
-   [x] My changes do not introduce new security vulnerabilities
-   [x] Run `yarn lint:fix`

### CODE REVIEW

**Review by vivianjeng — CHANGES_REQUESTED** [review]
I encountered this error when posting
```
PostList.tsx:221 Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
    at Object.throwInvalidHookError (react-dom.development.js:16227:9)
    at useContext (react.development.js:1619:21)
    at useUser (User.tsx:312:42)
    at useActionCount (useActionCount.ts:46:26)
    at randomNonce (randomNonce.ts:14:52)
    at eval (useCreatePost.ts:63:47)
    at Generator.next (<anonymous>)
    at fulfilled (useCreatePost.ts:12:24)
```

please check

**Comment by kittybest** [comment]
I think the `randomNonce()` should also be called inside `createComment` hook?

File: packages/frontend/src/utils/randomNonce.ts
Comment (kittybest) [inline packages/frontend/src/utils/randomNonce.ts:1]: Could be changed to 
```ts
export default function randomNonce(actionCount: number) {
    if (actionCount < NUM_EPOCH_KEY_NONCE_PER_EPOCH) {
        return actionCount
    }

    return Math.floor(Math.random() * NUM_EPOCH_KEY_NONCE_PER_EPOCH)
}
```
This is only a logic function, no need to call any hook or context.

Comment (Xiawpohr) [reply packages/frontend/src/utils/randomNonce.ts:1]: Fixed

File: packages/frontend/src/config.ts [hunk 1]
```diff
@@ -1,2 +1,3 @@
 export const SERVER = process.env.SERVER ?? 'http://localhost:8000'
 export const KEY_SERVER = `${SERVER}/build/`
+export const NUM_EPOCH_KEY_NONCE_PER_EPOCH = 3
```
[the remark this answers is not in the document]
Comment (Xiawpohr) [reply packages/frontend/src/config.ts:3]: We can open an issue about this on Notion

[the remark this answers is not in the document]
Comment (Xiawpohr) [reply packages/frontend/src/config.ts:3]: Issue link: https://www.notion.so/Get-EpochKey-amount-per-epoch-from-api-config-ca956cae20554c2ebbf1aa7fd82f7750?pvs=4

**Comment by Xiawpohr** [comment]
@vivianjeng I already fixed this issue. I change the function name from `randomNonce` to `getEpochKeyNonce` and remove hook in it. Please review changes again. Thanks.

**Comment by vivianjeng** [comment]
截圖 2024-05-07 上午10 49 23

I still get the same epoch key at the 2nd and 3rd post.

**Comment by vivianjeng** [comment]
I printed out the action count and the nonce, but the behavior is weird. please check
```js
// post 1
action count 0
useCreatePost.ts:81 nonce 2
// post 2
useCreatePost.ts:79 action count 0
useCreatePost.ts:81 nonce 1
// post 3
useCreatePost.ts:79 action count 2
useCreatePost.ts:81 nonce 1
 
```

**Review by vivianjeng — APPROVED** [review]
need to change user state

**Pull request merged**